haystack/nodes/_json_schema.py

```diff
@@ -126,6 +126,39 @@ def find_subclasses_in_modules(importable_modules: List[str]):
     ]
 
 
+def handle_optional_params(param_fields: List[inspect.Parameter], params_schema: Dict[str, Any]) -> Dict[str, Any]:
+    """
+    Pydantic v1 cannot generate correct JSON schemas including Optional fields.
+    (https://github.com/samuelcolvin/pydantic/issues/1270)
+    This function detects optional parameters and updates the schema,
+    to allow null values for these parameters.
+    To be removed when Pydantic v2 is released and adopted
+    """
+    optional_params = []
+    for param in param_fields:
+        try:
+            if param.annotation.__origin__ == Union and type(None) in param.annotation.__args__:
+                optional_params.append(param)
+        except:
+            pass
+
+    for param in optional_params:
+        param_dict = params_schema["properties"][param.name]
+        type_ = param_dict.pop("type", None)
+        if type_:
+            if type_ == "array" and "items" in param_dict:
+                items = param_dict.pop("items")
+                param_dict["anyOf"] = [{"type": type_, "items": items}, {"type": "null"}]
```

Comment:
Technically lists and dicts can contain other optional parameters. This step might need to be made somehow recursive to work in the general case. However, given how rare the case is and the fact that this is a patch waiting for Pydantic v2, if you can't find an easy way to make it work, I think we can live with it.

Reply:
Being a patch, my implementation is tightly coupled with haystack schema and I didn't find these cases, even if surely possible.